<a href="https://colab.research.google.com/github/menguelx7/Diplodatos/blob/main/TP_ApNoSup.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Aprendizaje no Supervisado - Año 2022

# GRUPO 1- Trabajo práctico entregable N°1
**********************************************************************

Integrantes: 


*   Laura Ailin Feu Marquestau
*   Matias Miguel Mengual Figueroa 
*   Marcelo Perea
*   Natalia Giselle Tassin
*   Ignacio Villagra Torcomian

Práctico entregable: 
Utilizar la base de jugadores “players_22.csv” disponible en la página de Kaggle https://www.kaggle.com/datasets/stefanoleone992/fifa-22-complete-player-dataset . Considerar que la base 2022 no tiene el mismo formato que la base vista en clase, a los nombres de las variables se les agregó una keyword para identificar a qué tipo de habilidad corresponde.

Con la nueva base, realizar un análisis análogo al que realizamos en el cursado de la materia con los datos FIFA2019. Realice comentarios en cada parte (verbose=True ;))


1.   Análisis exploratorio de la base.
2.   Evaluación visual  e intuitiva de a dos variables numéricas por vez.
3.   Uso de dos técnicas de clustering: por ejemplo k-medias, DBSCAN, mezcla de Gaussianas y/o alguna jerárquica. Elección justificada de hiper-parámetros
4.   Evaluación y Análisis de los clusters encontrados.
5.   Pregunta: ¿Se realizó alguna normalización o escalado de la base? ¿Por qué ?
6.   Uso de alguna transformación (proyección, Embedding) para visualizar los resultados y/o usarla como preprocesado para aplicar alguna técnica de clustering.




# Cargamos las librerias y el DataSet a utilizar


In [ ]:
#Importamos las librerias que vamos a usar

import numpy as np
import pandas as pd
pd.set_option('display.max_columns',100)
pd.set_option('display.max_rows',1000)
import itertools
import warnings
warnings.filterwarnings("ignore")
import io
from plotly.offline import init_notebook_mode, plot,iplot
import plotly.graph_objs as go
init_notebook_mode(connected=True)
import matplotlib.pyplot as plt
import plotly.tools as tls#visualization
import plotly.figure_factory as ff#visualization
import seaborn as sns

In [ ]:
#cargamos el DataSet
df = pd.read_csv('/content/players_22.csv') 
df.head(3)

sofifa_id                                         player_url  \
0     158023  https://sofifa.com/player/158023/lionel-messi/...   
1     188545  https://sofifa.com/player/188545/robert-lewand...   
2      20801  https://sofifa.com/player/20801/c-ronaldo-dos-...   

          short_name                            long_name player_positions  \
0           L. Messi       Lionel Andrés Messi Cuccittini       RW, ST, CF   
1     R. Lewandowski                   Robert Lewandowski               ST   
2  Cristiano Ronaldo  Cristiano Ronaldo dos Santos Aveiro           ST, LW   

   overall  potential    value_eur  wage_eur  age         dob  height_cm  \
0       93         93   78000000.0  320000.0   34  1987-06-24        170   
1       92         92  119500000.0  270000.0   32  1988-08-21        185   
2       91         91   45000000.0  270000.0   36  1985-02-05        187   

   weight_kg  club_team_id            club_name             league_name  \
0         72          73.0  Paris Saint-Germain          French Ligue 1   
1         81          21.0    FC Bayern München    German 1. Bundesliga   
2         83          11.0    Manchester United  English Premier League   

   league_level club_position  club_jersey_number club_loaned_from  \
0           1.0            RW                30.0              NaN   
1           1.0            ST                 9.0              NaN   
2           1.0            ST                 7.0              NaN   

  club_joined  club_contract_valid_until  nationality_id nationality_name  \
0  2021-08-10                     2023.0              52        Argentina   
1  2014-07-01                     2023.0              37           Poland   
2  2021-08-27                     2023.0              38         Portugal   

   nation_team_id nation_position  nation_jersey_number preferred_foot  \
0          1369.0              RW                  10.0           Left   
1          1353.0              RS                   9.0          Right   
2          1354.0              ST                   7.0          Right   

   weak_foot  skill_moves  international_reputation    work_rate body_type  \
0          4            4                         5   Medium/Low    Unique   
1          4            4                         5  High/Medium    Unique   
2          4            5                         5     High/Low    Unique   

  real_face  release_clause_eur  \
0       Yes         144300000.0   
1       Yes         197200000.0   
2       Yes          83300000.0   

                                         player_tags  \
0  #Dribbler, #Distance Shooter, #FK Specialist, ...   
1  #Aerial Threat, #Distance Shooter, #Clinical F...   
2  #Aerial Threat, #Dribbler, #Distance Shooter, ...   

                                       player_traits  pace  shooting  passing  \
0  Finesse Shot, Long Shot Taker (AI), Playmaker ...  85.0      92.0     91.0   
1  Solid Player, Finesse Shot, Outside Foot Shot,...  78.0      92.0     79.0   
2  Power Free-Kick, Flair, Long Shot Taker (AI), ...  87.0      94.0     80.0   

   dribbling  defending  physic  attacking_crossing  attacking_finishing  \
0       95.0       34.0    65.0                  85                   95   
1       86.0       44.0    82.0                  71                   95   
2       88.0       34.0    75.0                  87                   95   

   attacking_heading_accuracy  attacking_short_passing  attacking_volleys  \
0                          70                       91                 88   
1                          90                       85                 89   
2                          90                       80                 86   

   skill_dribbling  skill_curve  ...  power_stamina  power_strength  \
0               96           93  ...             72              69   
1               85           79  ...             76              86   
2               88           81  ...             77              77   

   power_long_shots  ment

In [ ]:
df.tail(3)

sofifa_id                                         player_url  \
19236     262760  https://sofifa.com/player/262760/nathan-logue/...   
19237     262820  https://sofifa.com/player/262820/luke-rudden/2...   
19238     264540  https://sofifa.com/player/264540/emanuel-lalch...   

               short_name                long_name player_positions  overall  \
19236            N. Logue  Nathan Logue-Cunningham               CM       47   
19237           L. Rudden              Luke Rudden               ST       47   
19238  E. Lalchhanchhuaha  Emanuel Lalchhanchhuaha              CAM       47   

       potential  value_eur  wage_eur  age         dob  height_cm  weight_kg  \
19236         55   100000.0     500.0   21  2000-02-07        178         72   
19237         60   110000.0     500.0   19  2002-04-23        173         66   
19238         60   110000.0     500.0   19  2002-02-06        167         61   

       club_team_id            club_name                     league_name  \
19236      111131.0           Finn Harps  Rep. Ireland Airtricity League   
19237      111131.0           Finn Harps  Rep. Ireland Airtricity League   
19238      113040.0  NorthEast United FC             Indian Super League   

       league_level club_position  club_jersey_number club_loaned_from  \
19236           1.0           RES                31.0              NaN   
19237           1.0           RES                12.0              NaN   
19238           1.0           SUB                13.0              NaN   

      club_joined  club_contract_valid_until  nationality_id  \
19236  2021-06-18                     2021.0              25   
19237  2021-06-18                     2021.0              25   
19238  2021-08-31                     2025.0             159   

          nationality_name  nation_team_id nation_position  \
19236  Republic of Ireland             NaN             NaN   
19237  Republic of Ireland             NaN             NaN   
19238                India             NaN             NaN   

       nation_jersey_number preferred_foot  weak_foot  skill_moves  \
19236                   NaN          Right          3            2   
19237                   NaN          Right          3            2   
19238                   NaN          Right          3            2   

       international_reputation      work_rate       body_type real_face  \
19236                         1  Medium/Medium  Lean (170-185)        No   
19237                         1  Medium/Medium  Lean (170-185)        No   
19238                         1  Medium/Medium     Lean (170-)        No   

       release_clause_eur player_tags player_traits  pace  shooting  passing  \
19236            175000.0         NaN           NaN  60.0      37.0     45.0   
19237            239000.0         NaN           NaN  68.0      46.0     36.0   
19238            217000.0         NaN           NaN  68.0      38.0     45.0   

       dribbling  defending  physic  attacking_crossing  attacking_finishing  \
19236       49.0       41.0    52.0                  39                   32   
19237       48.0       15.0    42.0                  29                   49   
19238       48.0       36.0    48.0                  39                   32   

       attacking_heading_accuracy  attacking_short_passing  attacking_volleys  \
19236                          43                       49                 37   
19237                          40                       38                 34   
19238                          46                       50                 37   

       skill_dribbling  skill_curve  ...  power_stamina  power_strength  \
19236               47           37  ...             56              50   
19237               42           36  ...             50              40   
19238               48           38  ...             57              40   

       power_long_shots  mentality_aggression  mentality_interceptions  \
19236                34                    5

# 1. Análisis exploratorio de la base.

In [ ]:
#vemos el nombre las columnas que compónen nuestro DF
df.columns

Index(['sofifa_id', 'player_url', 'short_name', 'long_name',
       'player_positions', 'overall', 'potential', 'value_eur', 'wage_eur',
       'age',
       ...
       'lcb', 'cb', 'rcb', 'rb', 'gk', 'player_face_url', 'club_logo_url',
       'club_flag_url', 'nation_logo_url', 'nation_flag_url'],
      dtype='object', length=110)

In [ ]:
#Realizamos un describe 
df.describe().T

count          mean           std      min  \
sofifa_id                    19239.0  2.314681e+05  2.703972e+04     41.0   
overall                      19239.0  6.577218e+01  6.880232e+00     47.0   
potential                    19239.0  7.107937e+01  6.086213e+00     49.0   
value_eur                    19165.0  2.850452e+06  7.613700e+06   9000.0   
wage_eur                     19178.0  9.017989e+03  1.947018e+04    500.0   
age                          19239.0  2.521082e+01  4.748235e+00     16.0   
height_cm                    19239.0  1.812997e+02  6.863179e+00    155.0   
weight_kg                    19239.0  7.494303e+01  7.069434e+00     49.0   
club_team_id                 19178.0  5.058050e+04  5.440187e+04      1.0   
league_level                 19178.0  1.354364e+00  7.478651e-01      1.0   
club_jersey_number           19178.0  2.094525e+01  1.790937e+01      1.0   
club_contract_valid_until    19178.0  2.022764e+03  1.213203e+00   2021.0   
nationality_id               19239.0  5.860268e+01  5.029861e+01      1.0   
nation_team_id                 759.0  1.448085e+04  3.532873e+04   1318.0   
nation_jersey_number           759.0  1.256785e+01  7.039116e+00      1.0   
weak_foot                    19239.0  2.946151e+00  6.715605e-01      1.0   
skill_moves                  19239.0  2.352461e+00  7.676590e-01      1.0   
international_reputation     19239.0  1.094184e+00  3.710982e-01      1.0   
release_clause_eur           18063.0  5.374044e+06  1.494837e+07  16000.0   
pace                         17107.0  6.821307e+01  1.093315e+01     28.0   
shooting                     17107.0  5.234530e+01  1.405162e+01     18.0   
passing                      17107.0  5.731256e+01  1.006897e+01     25.0   
dribbling                    17107.0  6.256117e+01  9.651312e+00     27.0   
defending                    17107.0  5.170363e+01  1.618975e+01     14.0   
physic                       17107.0  6.482329e+01  9.791886e+00     29.0   
attacking_crossing           19239.0  4.957742e+01  1.803466e+01      6.0   
attacking_finishing          19239.0  4.589443e+01  1.972102e+01      2.0   
attacking_heading_accuracy   19239.0  5.178388e+01  1.729418e+01      5.0   
attacking_short_passing      19239.0  5.886798e+01  1.449086e+01      7.0   
attacking_volleys            19239.0  4.246385e+01  1.765333e+01      3.0   
skill_dribbling              19239.0  5.566043e+01  1.878459e+01      4.0   
skill_curve                  19239.0  4.726893e+01  1.818109e+01      6.0   
skill_fk_accuracy            19239.0  4.224903e+01  1.717859e+01      4.0   
skill_long_passing           19239.0  5.307225e+01  1.502657e+01      9.0   
skill_ball_control           19239.0  5.847201e+01  1.666372e+01      8.0   
movement_acceleration        19239.0  6.465289e+01  1.516740e+01     14.0   
movement_sprint_speed        19239.0  6.471490e+01  1.496543e+01     15.0   
movement_agility             19239.0  6.350008e+01  1.486228e+01     18.0   
movement_reactions           19239.0  6.145002e+01  9.042281e+00     25.0   
movement_balance             19239.0  6.406861e+01  1.432479e+01     15.0   
power_shot_power             19239.0  5.777686e+01  1.319222e+01     20.0   
power_jumping                19239.0  6.481350e+01  1.212298e+01     22.0   
power_stamina                19239.0  6.308488e+01  1.614528e+01     12.0   
power_strength               19239.0  6.500774e+01  1.266352e+01     19.0   
power_long_shots             19239.0  4.664270e+01  1.941158e+01      4.0   
mentality_aggression         19239.0  5.553896e+01  1.697218e+01     10.0   
mentality_interceptions      19239.0  4.661355e+01  2.067708e+01      3.0   
mentality_positioning        19239.0  5.033021e+01  1.962160e+01      2.0   
mentality_vision             19239.0  5.396460e+01  1.365048e+01     10.0   
mentality_penalties          19239.0  4.785872e+01  1.576858e+01      7.0   
mentality_composure          19239.0  5.792983e+01  1.215933e+01     12.0   
defending_marking_awareness

In [ ]:
from pandas.api.types import is_numeric_dtype

def info_univariado(df):
  df_info = pd.DataFrame(columns=['Cantidad', 'Tipo' , 'Missing', 'Unicos', 'Numeric'])
  for col in df:
    data_series = df[col]
    df_info.loc[col] = [data_series.count(), data_series.dtype, data_series.isnull().sum(), data_series.nunique(), is_numeric_dtype(data_series)]
    df_describe = df.describe(include='all').T[['top', 'mean', 'std', 'min', '25%', '50%', '75%', 'max']]
    df_stats = pd.DataFrame([df.skew(), df.kurtosis()], index=['sesgo', 'kurt']).T
  return pd.concat([df_info,pd.concat([df_describe, df_stats], axis=1)], axis=1).fillna('0')

df_uni_stats = info_univariado(df)
df_uni_stats

Cantidad     Tipo  Missing  Unicos  Numeric  \
sofifa_id                       19239    int64        0   19239     True   
player_url                      19239   object        0   19239    False   
short_name                      19239   object        0   18145    False   
long_name                       19239   object        0   19219    False   
player_positions                19239   object        0     674    False   
overall                         19239    int64        0      47     True   
potential                       19239    int64        0      46     True   
value_eur                       19165  float64       74     252     True   
wage_eur                        19178  float64       61     133     True   
age                             19239    int64        0      29     True   
dob                             19239   object        0    6312    False   
height_cm                       19239    int64        0      49     True   
weight_kg                       19239    int64        0      58     True   
club_team_id                    19178  float64       61     701     True   
club_name                       19178   object       61     701    False   
league_name                     19178   object       61      55    False   
league_level                    19178  float64       61       5     True   
club_position                   19178   object       61      29    False   
club_jersey_number              19178  float64       61      99     True   
club_loaned_from                 1102   object    18137     334    False   
club_joined                     18076   object     1163    1845    False   
club_contract_valid_until       19178  float64       61       9     True   
nationality_id                  19239    int64        0     163     True   
nationality_name                19239   object        0     163    False   
nation_team_id                    759  float64    18480      33     True   
nation_position                   759   object    18480      28    False   
nation_jersey_number              759  float64    18480      28     True   
preferred_foot                  19239   object        0       2    False   
weak_foot                       19239    int64        0       5     True   
skill_moves                     19239    int64        0       5     True   
international_reputation        19239    int64        0       5     True   
work_rate                       19239   object        0       9    False   
body_type                       19239   object        0      10    False   
real_face                       19239   object        0       2    False   
release_clause_eur              18063  float64     1176    1235     True   
player_tags                      1441   object    17798      71    False   
player_traits                    9398   object     9841    1051    False   
pace                            17107  float64     2132      70     True   
shooting                        17107  float64     2132      76     True   
passing                         17107  float64     2132      67     True   
dribbling                       17107  float64     2132      68     True   
defending                       17107  float64     2132      77     True   
physic                          17107  float64     2132      61     True   
attacking_crossing              19239    int64        0      88     True   
attacking_finishing             19239    int64        0      94     True   
attacking_heading_accuracy      19239    int64        0      89     True   
attacking_short_passing         19239    int64        0      86     True   
attacking_volleys               19239    int64        0      88     True   
skill_dribbling                 19239    int64        0      92     True   
skill_curve                     19239    int64        0      89     True   
skill_fk_accuracy               19239    int64        0      90     True   
skill_long_passing              19239    int64        0      85     True   
s

In [ ]:
#Corroboramos si tenemos datos duplicados
df.duplicated().sum()

0

Podemos corroborar que nuestro DataFrame no contiene datos duplicados

In [ ]:
#Con el shape corroboramos cuantas filas y columnas tenemos en nuestro DF
df.shape

(19239, 110)

Nuestro DataFrame tiene:
*   19239 Filas
*   110 Columnas



In [ ]:
#Corroboramos que tipo de dato corresponde a cada columna
df.dtypes

sofifa_id                        int64
player_url                      object
short_name                      object
long_name                       object
player_positions                object
overall                          int64
potential                        int64
value_eur                      float64
wage_eur                       float64
age                              int64
dob                             object
height_cm                        int64
weight_kg                        int64
club_team_id                   float64
club_name                       object
league_name                     object
league_level                   float64
club_position                   object
club_jersey_number             float64
club_loaned_from                object
club_joined                     object
club_contract_valid_until      float64
nationality_id                   int64
nationality_name                object
nation_team_id                 float64
nation_position          

Los tipos de datos que contiene nuestro DataFrame son:
*   int64
*   object
*   float64



In [ ]:
#Vamos a sumar los nulos en todo el DF
df.isnull().sum()

sofifa_id                          0
player_url                         0
short_name                         0
long_name                          0
player_positions                   0
overall                            0
potential                          0
value_eur                         74
wage_eur                          61
age                                0
dob                                0
height_cm                          0
weight_kg                          0
club_team_id                      61
club_name                         61
league_name                       61
league_level                      61
club_position                     61
club_jersey_number                61
club_loaned_from               18137
club_joined                     1163
club_contract_valid_until         61
nationality_id                     0
nationality_name                   0
nation_team_id                 18480
nation_position                18480
nation_jersey_number           18480
p

Podemos validar que de las 110 columnas que tiene nuestro DataFrame 27 de ellas contienen datos nulos, los cuales son:

*   value_eur con 74 nulls
*   wage_eur con 64 nulls
*   club_team_id con 61 nulls
*   club_name con 61 nulls
*   league_name con 61 nulls
*   league_level con 61 nulls
*   club_position con 61 nulls
*   club_jersey_number con 61 nulls
*   club_loaned_from con 18137 nulls
*   club_joined con 1163 nulls
*   club_contract_valid_until con 61 nulls
*   nation_team_id con 18480 nulls
*   nation_position con 18480 nulls
*   nation_jersey_number con 18480 nulls
*   release_clause_eur con 1176 nulls
*   player_tags con 17798 nulls
*   player_traits con 9841 nulls 
*   pace con 2132 nulls
*   shooting con 2132 nulls
*   passing con 2132 nulls
*   dribbling con 2132 nulls
*   defending con 2132 nulls
*   physic con 2132 nulls
*   goalkeeping_speed con 17107 nulls
*   club_logo_url con 61 nulls
*   club_flag_url con 61 nulls
*   nation_logo_url con 18480 nulls

In [ ]:
#Ahora veamos los datos unicos delDF
df.nunique()

sofifa_id                      19239
player_url                     19239
short_name                     18145
long_name                      19219
player_positions                 674
overall                           47
potential                         46
value_eur                        252
wage_eur                         133
age                               29
dob                             6312
height_cm                         49
weight_kg                         58
club_team_id                     701
club_name                        701
league_name                       55
league_level                       5
club_position                     29
club_jersey_number                99
club_loaned_from                 334
club_joined                     1845
club_contract_valid_until          9
nationality_id                   163
nationality_name                 163
nation_team_id                    33
nation_position                   28
nation_jersey_number              28
p

# 2. Evaluación visual e intuitiva de a dos variables numéricas por vez.